### ISSUE TITLE
Syntax and other errors when using the Cpp runtime of ANTLR4 with grammars generated with antlr4.13.2-complete.jar

### ISSUE THREAD

**Issue by peytonk132** [antlr/antlr4#4674]
Hello everyone, I've been trying to create a parser using antlr in VS 2022 for awhile now but I can never seem to shake these syntax errors. I had originally posted on Stackoverflow as requested but I didn't really get any answers there even though the people were very helpful. I had originally generated grammar files from https://github.com/antlr/grammars-v4/tree/master/fortran/fortran90 for the Cpp targets and in my folly hadn't used the transformGrammars.py file but had then used them after realizing my mistake and still got the errors. They range from various things like certain classes not being base class members to general syntax errors. The one file that i think holds the key to fixing a lot of other errors is ParseTreeType.h. These are my errors for for that file:
```
ParseTreeType.h(20,13): error C2143: syntax error: missing '}' before 'constant'
ParseTreeType.h(20,13): error C2059: syntax error: 'constant'
ParseTreeType.h(22,9): error C2143: syntax error: missing ';' before '}'
ParseTreeType.h(25,1): error C2059: syntax error: '}'
ParseTreeType.h(25,1): error C2143: syntax error: missing ';' before '}'
```

and this is the file I have:
```
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

#pragma once

#include <cstddef>

#include "../mainfunc/antlr-files/antlr4-common.h"

namespace antlr4
{
    namespace tree
    {

        enum class ParseTreeType : size_t
        {
            TERMINAL = 1,
            ERROR = 2,
            RULE = 3,
        };

    } // namespace tree
} // namespace antlr4
```

Has anybody else experienced this before? Also, almost all the include files in the runtime src files were completely wack and didn't work at all. Could this be a bug or am I just stupid?

**Comment by kaby76** [comment]
Did you follow my comments to [first build the runtime](https://stackoverflow.com/questions/78856543/over-100-compiler-errors-when-using-antlr-4-13-2-with-the-c-runtime#comment139035040_78856543) then [build the app that I wrote for you](https://stackoverflow.com/questions/78856543/over-100-compiler-errors-when-using-antlr-4-13-2-with-the-c-runtime#comment139035159_78856543)?

The grammar works for Cpp on Windows with VS2022.
```
PS C:\msys64\home\Kenne\fortran90-Cpp> msbuild .\MyParser.sln -p:Configuration=Debug > out.txt
PS C:\msys64\home\Kenne\fortran90-Cpp> .\x64\Debug\MyParser.exe C:\msys64\home\Kenne\issues\g4-current\fortran\fortran90\examples\flang\test\f90_correct\src\ac02.f90
Cpp 0 C:\msys64\home\Kenne\issues\g4-current\fortran\fortran90\examples\flang\test\f90_correct\src\ac02.f90 success 0.688818
Total Time: 0.729936
PS C:\msys64\home\Kenne\fortran90-Cpp>
```

As I mentioned in SO, you may have a `#define` that is conflicting with the runtime. But, you need to first verify that your environment is at least set up correctly.

**Comment by kaby76** [comment]
I found the project you are working on here: https://github.com/peytonk132/FortIDE. Please check in the code you are working on (preferably in a branch of main `git clone https://github.com/peytonk132/FortIDE.git; cd FortIDE; git branch with-antlr; git checkout with-antlr; ...... (modify with your new code) ...; (add, commit, and push changes).`. Then, close this issue (https://github.com/antlr/antlr4/issues/4674). I can help you get this compiling.

Basically, you should not be modifying the Antlr4 sources. You need to change the .vcxproj to add in an include path for every library you are referencing. It's likely that you also aren't specifying the language level correctly, among other things.

**Comment by kaby76** [comment]
Just in case someone happens to see the same compilation errors (e.g., `ParseTreeType.h(17,5): error C2143: syntax error: missing '}'
 before 'constant'` or `IntStream.h(30,64): error C2589: '(': illegal token on right side
of '::'`), it is caused by `#define` conflicts when including `windows.h` before Antlr4 runtime headers. One way to fix this is to silo the code for windows from the code that uses antlr4. But, it may just as well be solved by reversing the order of the `#include`. Such as life with C++ and the preprocessor.

**Comment by wuxianggujun** [comment]
> Just in case someone happens to see the same compilation errors (e.g., `ParseTreeType.h(17,5): error C2143: syntax error: missing '}' before 'constant'` or `IntStream.h(30,64): error C2589: '(': illegal token on right side of '::'`), it is caused by `#define` conflicts when including `windows.h` before Antlr4 runtime headers. One way to fix this is to silo the code for windows from the code that uses antlr4. But, it may just as well be solved by reversing the order of the `#include`. In addition, I found (at least for the driver for the Cpp target used in grammars-v4) that after `#include <windows.h>`, I also needed to add `#undef min`. Apparently, `min()` is defined as a macro. Such as life with C++ and the preprocessor.以防万一有人碰巧看到相同的编译错误（例如， `ParseTreeType.h(17,5): error C2143: syntax error: missing '}' before 'constant'` 或者 `IntStream.h(30,64): error C2589: '(': illegal token on right side of '::'` ），这是由于在 Antlr4 运行时标头之前包含`windows.h`时`#define`冲突引起的。解决此问题的一种方法是将 Windows 代码与使用 antlr4 的代码隔离开来。但是，也可以通过颠倒`#include`的顺序来解决。另外，我发现（至少对于在 Grammars-v4 中使用的 Cpp 目标的驱动程序）在`#include <windows.h>`之后，我还需要添加`#undef min` 。显然， `min()`被定义为一个宏。例如 C++ 和预处理器的生活。

I checked some information and found that it was a macro definition conflict, but my project is a qt project and a lot of third-party libraries reference the window header file, and then I defined #define NOMINMAX in the cmake configuration, but still got a lot of errors. Do you have any solution?

**Comment by wuxianggujun** [comment]
> > Just in case someone happens to see the same compilation errors (e.g., `ParseTreeType.h(17,5): error C2143: syntax error: missing '}' before 'constant'` or `IntStream.h(30,64): error C2589: '(': illegal token on right side of '::'`), it is caused by `#define` conflicts when including `windows.h` before Antlr4 runtime headers. One way to fix this is to silo the code for windows from the code that uses antlr4. But, it may just as well be solved by reversing the order of the `#include`. In addition, I found (at least for the driver for the Cpp target used in grammars-v4) that after `#include <windows.h>`, I also needed to add `#undef min`. Apparently, `min()` is defined as a macro. Such as life with C++ and the preprocessor.以防万一有人碰巧看到相同的编译错误（例如， `ParseTreeType.h(17,5): error C2143: syntax error: missing '}' before 'constant'` 或者 `IntStream.h(30,64): error C2589: '(': illegal token on right side of '::'` ），这是由于在 Antlr4 运行时标头之前包含`windows.h`时`#define`冲突引起的。解决此问题的一种方法是将 Windows 代码与使用 antlr4 的代码隔离开来。但是，也可以通过颠倒`#include`的顺序来解决。另外，我发现（至少对于在 Grammars-v4 中使用的 Cpp 目标的驱动程序）在`#include <windows.h>`之后，我还需要添加`#undef min` 。显然， `min()`被定义为一个宏。例如 C++ 和预处理器的生活。以防万一有人碰巧看到相同的编译错误（例如， `ParseTreeType.h(17,5): error C2143: syntax error: missing '}' before 'constant'` 或者 `IntStream.h(30,64): error C2589: '(': illegal token on right side of '::'` ），这是由于在 Antlr4 运行时标头之前包含`windows.h`时`#define`冲突引起的。解决此问题的一种方法是将 Windows 代码与使用 antlr4 的代码隔离开来。但是，也可以通过颠倒`#include`的顺序来解决。另外，我发现（至少对于在 Grammars-v4 中使用的 Cpp 目标的驱动程序）在`#include <windows.h>`之后，我还需要添加`#undef min` 。显然， `min()`被定义为一个宏。比如 C++ 和预处理器的生活。万一有人碰巧看到相同的编译错误（例如， `ParseTreeType.h(17,5): error C2143: syntax error: missing '}' before 'constant'` 或者 `IntStream.h(30,64): error C2589: '(': illegal token on right side of '::'` ），这是由于在 Antlr4 运行时标头之前包含`windows.h`时`#define`冲突引起的。解决此问题的一种方法可以将 Windows 代码与使用 antlr4 的代码隔离开来。但是，也可以通过颠倒`#include`另外，我发现`#include`至少对于在 Grammars-v4 中使用的 Cpp 目标的驱动程序）在`#include <windows.h>`之后，我还需要添加`#undef min` 。显然， `min()`被定义为一个宏。例如 C++ 和损耗器的生活。
> 
> I checked some information and found that it was a macro definition conflict, but my project is a qt project and a lot of third-party libraries reference the window header file, and then I defined #define NOMINMAX in the cmake configuration, but still got a lot of errors. Do you have any solution?我查了一些资料，发现是宏定义冲突，但是我的项目是qt项目，很多第三方库引用了window头文件，然后我在cmake配置中定义了#define NOMINMAX，但是还是得到了很多错误。你有什么解决办法吗？

// 在头文件开始处
#pragma push_macro("ERROR")
#pragma push_macro("emit")
#undef ERROR
#undef emit

// 你的代码

// 在头文件结束处
#pragma pop_macro("ERROR")
#pragma pop_macro("emit")
